renderdoc/driver/vulkan/vk_resources.h

```diff
@@ -1071,6 +1071,47 @@ struct MemRefs
 struct ImgRefs;
 struct ImageState;
 
+class VkPendingSubmissionCompleteCallbacks
+{
+public:
+  VkPendingSubmissionCompleteCallbacks() = default;
+  VkPendingSubmissionCompleteCallbacks(VkDevice device, VkEvent event,
+                                       rdcarray<std::function<void()>> callbacks);
+  VkPendingSubmissionCompleteCallbacks(const VkPendingSubmissionCompleteCallbacks &other);
+  VkPendingSubmissionCompleteCallbacks(VkPendingSubmissionCompleteCallbacks &&other);
+  VkPendingSubmissionCompleteCallbacks &operator=(VkPendingSubmissionCompleteCallbacks other);
+
+  ~VkPendingSubmissionCompleteCallbacks();
+
+  operator VkEvent() const
+  {
+    SCOPED_LOCK(m_Lock);
+    return d == NULL ? VK_NULL_HANDLE : d->event;
+  }
+
+  void swap(VkPendingSubmissionCompleteCallbacks &other)
+  {
+    using std::swap;
+
+    swap(callbacks, other.callbacks);
+    swap(d, other.d);
+  }
+
+  rdcarray<std::function<void()>> callbacks;
+
+private:
+  struct DPtr
+  {
+    int32_t refCount;
+    VkDevice device;
+    VkEvent event;
+  };
+
+  static Threading::CriticalSection m_Lock;
```

Comment:
Having a global lock here across all objects is not the best way to implement refcounting as it will introduce contention between unrelated objects and is a heavy-weight solution to synchronisation. If you look at how `ResourceRecord` is implemented with the code snippet I posted before it uses the typical approach of atomic operations to count the refcount and delete when it reaches 0.